### PR TITLE
Fix non-deterministic caching of ProjectDescriptionHelpers

### PR DESCRIPTION
**Pull request by Ernest0-Production**
Resolves https://github.com/tuist/tuist/issues/6374

### Short description 📝

Fixed several issues related to the `ProjectDescriptionHelpers` build and cache:

1. `ProjectDescriptionHelpersBuilder` was not shared between manifests, but was created anew for each manifest. 

This made it pointless to have a `ProjectDescriptionHelpersBuilder.buildHelpers` dictionary to cache already built helper modules.

I created `CachedProjectDescriptionHelpersBuilderFactory` decorator to solve this problem:
```swift
class CachedProjectDescriptionHelpersBuilderFactory: ProjectDescriptionHelpersBuilderFactoring {
    private let builderFactory: ProjectDescriptionHelpersBuilderFactoring
    private var helperBuildersCache: ThreadSafe<[AbsolutePath: ProjectDescriptionHelpersBuilding]> = ThreadSafe([:])

   func projectDescriptionHelpersBuilder(cacheDirectory: AbsolutePath) -> any ProjectDescriptionHelpersBuilding {
        return helperBuildersCache.mutate { builders in
            if let helpersBuilder = builders[cacheDirectory] {
                return helpersBuilder
            } else {
                let newBuilder = builderFactory.projectDescriptionHelpersBuilder(cacheDirectory: cacheDirectory)
                builders[cacheDirectory] = newBuilder
                return newBuilder
            }
        }
    }
}
```

2. Fixed a race condition in shared* `ProjectDescriptionHelpersBuilder` when building a module in parallel from several threads (for each manifests).

@pepicrft  recently wrapped `ProjectDescriptionHelpersBuilder.buildHelpers` in `ThreadSafe`, but this did **not solve the race condition problem correctly**.

- It was necessary to create a `ProjectDescriptionHelpersModule` or retrieve it from the `.buildHelpers` cache in **one atomic operation**
- It was necessary to assemble the helpers module or check its presence in `cacheDirectory` and compilation as part of **another atomic operation** tied to the passed `path`.

3. Removed the meaningless `ProjectDescriptionHelpersHashing.prefixHash` when generating the path for the `ProjectDescriptionHelpers` dyld , since the prefix was **vulnerable to collision** due to the fact that part of the hash is cut off at 7 characters(some kind of magic number):
```
computeHash(for: "Users/jenkins/workspace/PR-2962"), // 4be35b3
computeHash(for: "Users/jenkins/workspace/PR-5162") // 4be35b3
```

### How to test the changes locally 🧐

I tested everything on my working project and set `print` in Tuist Sources to make sure that the ProjectDescriptionHelpers module was built only once.

**I can't reliably reproduce the race condition issue because it requires a lot of manifests and a lot of iterations.**

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

**Comment by Ernest0-Production** [comment]
@fortmarek Hi!)

I finally found some free time and was able to figure out how to stabilize caching of `ProjectDescriptionHelpers` module.

Can you please validate my solution? 🙏

File: Sources/TuistLoader/ProjectDescriptionHelpers/CachedProjectDescriptionHelpersBuilderFactory.swift
Comment (fortmarek) [inline Sources/TuistLoader/ProjectDescriptionHelpers/CachedProjectDescriptionHelpersBuilderFactory.swift:8]: I don't think we ever use the `ProjectDescriptionHelpersBuilderFactory` by itself – should we cache the builder directly in the `ProjectDescriptionHelpersBuilderFactory` since we don't need the distinction between cached and non-cached factory?

File: Sources/TuistLoader/ProjectDescriptionHelpers/CachedProjectDescriptionHelpersBuilderFactory.swift
Comment (fortmarek) [inline Sources/TuistLoader/ProjectDescriptionHelpers/CachedProjectDescriptionHelpersBuilderFactory.swift:11]: Do we actually need an extra level of caching? Isn't the underlying problem we _always_ create a new `ProjectDescriptionsBuilder` [here](https://github.com/tuist/tuist/blob/main/Sources/TuistLoader/Loaders/ManifestLoader.swift#L380)? The `cacheDirectory` is the same across runs, so we could even create a single `ProjectDescriptionHelpersBuilder` in the init of `ManifestLoader` or have a `lazy var` in the `ManifestLoader`.

File: Sources/TuistLoader/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilder.swift
Comment (fortmarek) [inline Sources/TuistLoader/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilder.swift:190]: I don't know how likely it is to happen, but if helpers are currently being built, doesn't this mean we'll rebuild them again, unnecessarily? Should we instead save the project description helpers builder process using something like `Task` and wait for the result instead of starting the process again?

**Review by fortmarek — COMMENTED** [review]
These are amazing findings! A couple of questions around the implementation, but overall, the direction of this PR makes sense to me 👍

File: Sources/TuistLoader/ProjectDescriptionHelpers/CachedProjectDescriptionHelpersBuilderFactory.swift
Comment (Ernest0-Production) [reply Sources/TuistLoader/ProjectDescriptionHelpers/CachedProjectDescriptionHelpersBuilderFactory.swift:8]: You are right. This may look like an overhead
I did it on the same principle as it was done with `CachedManifestLoader` and `CachedContentHasher`

I can move the current logic to `ProjectDescriptionHelpersBuilderFactory` if you see fit

Comment (Ernest0-Production) [reply Sources/TuistLoader/ProjectDescriptionHelpers/CachedProjectDescriptionHelpersBuilderFactory.swift:8]: You are right. This may look like an overhead
I did it on the same principle as it was done with CachedManifestLoader and CachedContentHasher

I move the current logic to ProjectDescriptionHelpersBuilderFactory if you see fit

File: Sources/TuistLoader/ProjectDescriptionHelpers/CachedProjectDescriptionHelpersBuilderFactory.swift
Comment (Ernest0-Production) [reply Sources/TuistLoader/ProjectDescriptionHelpers/CachedProjectDescriptionHelpersBuilderFactory.swift:11]: > we could even create a single ProjectDescriptionHelpersBuilder in the init of ManifestLoader or have a lazy var 

This will be problematic, because... function for getting `cacheDirectory` from `cacheDirectoryProviderFactory` throwable:
- `throwing init` will lead to cascading changes in the project
- `lazy var` can not  be throwing

File: Sources/TuistLoader/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilder.swift
Comment (Ernest0-Production) [reply Sources/TuistLoader/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilder.swift:190]: > doesn't this mean we'll rebuild them again, unnecessarily? 
> starting the process again

You are right. This is possible if helper builder is initialized again.
And the reason is the `build.isBuilt` flag which is always false by default.

1. We can simply remove this flag:
I added it because... suggested that the module file might be created during the build process rather than when the command completes.

2. We can create a lock file in this directory to block re-building.
But I don’t yet know how to implement the process of waiting for the lock file to be deleted

Comment (Ernest0-Production) [reply Sources/TuistLoader/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilder.swift]: UPD: replace on Task

File: Sources/TuistLoader/ProjectDescriptionHelpers/CachedProjectDescriptionHelpersBuilderFactory.swift
Comment (Ernest0-Production) [reply Sources/TuistLoader/ProjectDescriptionHelpers/CachedProjectDescriptionHelpersBuilderFactory.swift]: Moved

**Review by fortmarek — COMMENTED** [review]
The implementation looks good to me, thanks a lot for incorporating the feedback!

What do you think of adding a new test case in [ProjectDescriptionHelpersBuilderIntegrationTests](https://github.com/tuist/tuist/blob/main/Tests/TuistLoaderIntegrationTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderIntegrationTests.swift) to ensure that if the same project description helpers are run in quick succession, they are built just one time?

You can either mock the system command or assert that the `Built \(name)...` log appears in the logger just once.

**Comment by Ernest0-Production** [comment]
It will be done) I'll do it soon

**Comment by Ernest0-Production** [comment]
@fortmarek Hi! Can you please re-check?

File: Tests/TuistLoaderTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderTests.swift
Comment (fortmarek) [inline Tests/TuistLoaderTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderTests.swift:51]: Is there a reason we can't stub the individual commands and instead we're allow some commands not to be stubbed?

**Review by fortmarek — APPROVED** [review]
Looks good. Just a minor comment around the stubs. Would like to wait for @pepicrft review on this one 👀

File: Tests/TuistLoaderTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderTests.swift
Comment (Ernest0-Production) [reply Tests/TuistLoaderTests/ProjectDescriptionHelpers/ProjectDescriptionHelpersBuilderTests.swift:51]: The command you need to stub looks veeeery complex. 
I went down this path at first, but then I thought that this test begins to check not only the behavior I need, but also the correctness of the command arguments, which makes it more fragile to refactoring.

**Comment by Ernest0-Production** [comment]
@pepicrft Hi! Can I ask you to take a look at this merge request so that it has already been merged? 🙏

**Comment by Ernest0-Production** [comment]
@fortmarek Hi! 👋 
Can someone else from the mainteners chedk this PR so that it can finally be merged :(

**Comment by fortmarek** [comment]
Hey @Ernest0-Production 👋 

Really sorry it's been taking so long. We'll do better from now on 🙏 I updated the branch with the latest changes from `main`, but it looks like the unit tests are failing. Would you mind taking a look? Once the CI is green, we can merge this.

**Comment by Ernest0-Production** [comment]
> Hey @Ernest0-Production 👋 
> 
> 
> 
> Really sorry it's been taking so long. We'll do better from now on 🙏 I updated the branch with the latest changes from `main`, but it looks like the unit tests are failing. Would you mind taking a look? Once the CI is green, we can merge this.

Hi! @fortmarek 
I'll try to do it over the weekend.

**Comment by Ernest0-Production** [comment]
@fortmarek I fixed merge conflicts.

**Review by fortmarek — APPROVED** [review]
Thanks a lot and sorry again for taking so long to review this one!

**Pull request merged**
